tests/flow/test_hnswlib.py

```diff
@@ -142,3 +142,90 @@ def test_recall_for_hnswlib_index_with_deletion():
+        vectors.append((i, vector))
+
+    # Create a random query vector and create a batch iterator
+    query_data = np.float32(rng.random((1, dim)))
+    batch_iterator = hnsw_index.create_batch_iterator(query_data)
+    labels_first_batch, distances_first_batch = batch_iterator.get_next_results(10, BY_ID)
+    for i, _ in enumerate(labels_first_batch[0][:-1]):
+        # Assert sorting by id
+        assert(labels_first_batch[0][i] < labels_first_batch[0][i+1])
+
+    labels_second_batch, distances_second_batch = batch_iterator.get_next_results(10, BY_SCORE)
+    for i, dist in enumerate(distances_second_batch[0][:-1]):
+        # Assert sorting by score
+        assert(distances_second_batch[0][i] < distances_second_batch[0][i+1])
+        # Assert that every distance in the second batch is higher than any distance of the first batch
+        assert(len(distances_first_batch[0][np.where(distances_first_batch[0] > dist)]) == 0)
+
+    # Reset
+    batch_iterator.reset()
+
+    # Run in batches of 100 until we reach 10000 results and measure recall
+    batch_size = 100
+    total_res = 1000
+    correct = 0
+    query_data = np.float32(rng.random((num_queries, dim)))
+    for target_vector in query_data:
+        batch_iterator = hnsw_index.create_batch_iterator(target_vector)
+        iterations = 0
+        # Sort distances of every vector from the target vector and get the actual order
+        dists = [(spatial.distance.euclidean(target_vector, vec), key) for key, vec in vectors]
+        dists = sorted(dists)
+        accumulated_labels = []
+        while batch_iterator.has_next():
+            iterations += 1
+            labels, distances = batch_iterator.get_next_results(batch_size, BY_SCORE)
+            accumulated_labels.extend(labels[0])
+            returned_results_num = len(accumulated_labels)
+            if returned_results_num == total_res:
+                returned_results_num = len(accumulated_labels)
+                keys = [key for _, key in dists[:returned_results_num]]
+                correct += len(set(accumulated_labels).intersection(set(keys)))
+                break
+        assert iterations == np.ceil(total_res/batch_size)
+    recall = float(correct) / (total_res*num_queries)
+    print(f'\nRecall for {total_res} results in index of size {num_elements} with dim={dim} is: ', recall)
+    assert recall > 0.9
+
+    # Run again a single query in batches until it is depleted.
+    batch_iterator = hnsw_index.create_batch_iterator(query_data[0])
+    iterations = 0
+    accumulated_labels = set()
+
+    while batch_iterator.has_next():
+        iterations += 1
+        labels, distances = batch_iterator.get_next_results(batch_size, BY_SCORE)
+        # Verify that we got new scores in each iteration.
+        assert len(accumulated_labels.intersection(set(labels[0]))) == 0
+        accumulated_labels = accumulated_labels.union(set(labels[0]))
+    assert len(accumulated_labels) >= 0.95*num_elements
```

Comment:
this is true always, no?

Reply:
Not necessarily, there might be a case where the number of nodes that are reachable from the entry point in the HNSW is small (we should choose the right `M` and `ef_construction` to avoid that...)